docs/eip/5559.md

```diff
@@ -0,0 +1,447 @@
+
+EIP-5559 is the first step toward such a tolerant CCIP-Write protocol which outlines how secure write deferrals can be made to L2s and centralised databases. The cases of L2s and databases are similar; deferral to an L2 involves routing the `eth_call` to another EVM, while deferral to a database can be made by extracting `eth_sign` from `eth_call` and posting the resulting signature explicitly along with the data for later verification.
+
+## Specification
+### Overview
+The following specification revolves around the structure and description of a cross-chain storage handler tasked with the responsibility of writing to an L2 or database storage. This draft introduces `StorageHandledByL2()` and `StorageHandledByDatabase()` storage handlers, and proposes that new `StorageHandledBy__()` reverts be allowed through new EIPs that sufficiently detail their interfaces and designs. Some foreseen examples of new storage handlers include `StorageHandledBySolana()` for Solana, `StorageHandledByFilecoin()` for Filecoin, `StorageHandledByIPFS()` for IPFS, `StorageHandledByIPNS()` for IPNS, `StorageHandledByArweave()` for Arweave, `StorageHandledByArNS()` for ArNS, `StorageHandledBySwarm()` for Swarm etc.
+
+![](https://raw.githubusercontent.com/namesys-eth/namesys-ccip-write/main/images/schematic.png)
+
+### L2 Handler: `StorageHandledByL2()`
+A minimal L2 handler only requires the list of `chainId` values and the corresponding `contract` addresses, while the clients must ensure that the calldata is invariant under routing to L2. One example implementation of an L2 handler in an L1 contract is shown below.
+
+```solidity
+// Define revert event
+error StorageHandledByL2(
+    address contractL2, 
+    uint256 chainId
+);
+
+// Generic function in a contract
+function setValue(
+    bytes32 node,
+    bytes32 key,
+    bytes32 value
+) external {
+    // Get metadata from on-chain sources
+    (
+        address contractL2, // Contract address on L2; may be globally constant
+        uint256 chainId // L2 ChainID; may be globally constant
+    ) = getMetadata(node); // Arbitrary code
+    // contract = 0x32f94e75cde5fa48b6469323742e6004d701409b
+    // chainId = 21
+    // Defer write call to L2 handler
+    revert StorageHandledByL2( 
+        contractL2,
+        chainId
+    );
+};
+```
+
+In this example, the deferral must prompt the client to build the transaction with the exact same original calldata, and submit it to the L2 by calling the exact same function on L2 as L1.
+
+```solidity
+// Function in L2 contract
+function setValue(
+    bytes32 node,
+    bytes32 key,
+    bytes32 value
+) external {
+    // Verify owner or manager permissions
+    require(authorised(node), "NOT_ALLOWED");
+    // Store data mapped by node & sender
+    data[keccak256(abi.encodePacked(node, msg.sender))][key] = value
```

Comment:
Does this line need to be specified as part of EIP?  `msg.sender` would be consistent across different chains for EOA addresses but not necessarily true for contract account

Reply:
It doesn't need to be specified. Wanted to put a very simple example for completeness but since it is faulty, perhaps better to leave it out.

Reply:
Fixed by https://github.com/ensdomains/docs/pull/224/commits/f50c32183c4aae26c06e7107f4ae8be9ff918601